File: src/redmine-client.ts
/**
 * Redmine REST API クライアント
 * Redmine API: https://www.redmine.org/projects/redmine/wiki/Rest_api
 */

export interface RedmineConfig {
  baseUrl: string;
  apiKey: string;
}

export interface Issue {
  id: number;
  project: { id: number; name: string };
  tracker: { id: number; name: string };
  status: { id: number; name: string };
  priority: { id: number; name: string };
  author: { id: number; name: string };
  assigned_to?: { id: number; name: string };
  subject: string;
  description?: string;
  start_date?: string;
  due_date?: string;
  done_ratio: number;
  created_on: string;
  updated_on: string;
}

export interface Project {
  id: number;
  name: string;
  identifier: string;
  description?: string;
  status: number;
  created_on: string;
  updated_on: string;
}

export interface IssueJournal {
  id: number;
  user: { id: number; name: string };
  notes: string;
  created_on: string;
}

export interface Attachment {
  id: number;
  filename: string;
  filesize: number;
  content_type: string;
  description?: string;
  content_url: string;
  author: { id: number; name: string };
  created_on: string;
}

export class RedmineClient {
  private baseUrl: string;
  private apiKey: string;

  constructor(config: RedmineConfig) {
    this.baseUrl = config.baseUrl.replace(/\/$/, "");
    this.apiKey = config.apiKey;
  }

  private async request<T>(path: string, options: RequestInit = {}): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    const response = await fetch(url, {
      ...options,
      headers: {
        "X-Redmine-API-Key": this.apiKey,
        "Content-Type": "application/json",
        ...options.headers,
      },
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Redmine API Error [${response.status}]: ${errorText}`);
    }

    return response.json() as Promise<T>;
  }

  // ─── Issues ──────────────────────────────────────────────

  async getIssues(params: {
    project_id?: string;
    status_id?: string;
    assigned_to_id?: string;
    subject?: string;
    limit?: number;
    offset?: number;
  }): Promise<{ issues: Issue[]; total_count: number }> {
    const query = new URLSearchParams();
    if (params.project_id) query.set("project_id", params.project_id);
    if (params.status_id) query.set("status_id", params.status_id);
    if (params.assigned_to_id) query.set("assigned_to_id", params.assigned_to_id);
    if (params.subject) query.set("subject", `~${params.subject}`);
    query.set("limit", String(params.limit ?? 25));
    query.set("offset", String(params.offset ?? 0));

    return this.request<{ issues: Issue[]; total_count: number }>(
      `/issues.json?${query.toString()}`
    );
  }

  async getIssue(id: number): Promise<{ issue: Issue & { journals: IssueJournal[]; attachments: Attachment[] } }> {
    return this.request(`/issues/${id}.json?include=journals,attachments`);
  }

  async createIssue(data: {
    project_id: string;
    subject: string;
    description?: string;
    tracker_id?: number;
    status_id?: number;
    priority_id?: number;
    assigned_to_id?: number;
    due_date?: string;
  }): Promise<{ issue: Issue }> {
    return this.request("/issues.json", {
      method: "POST",
      body: JSON.stringify({ issue: data }),
    });
  }

  async updateIssue(
    id: number,
    data: {
      subject?: string;
      description?: string;
      status_id?: number;
      priority_id?: number;
      assigned_to_id?: number;
      due_date?: string;
      done_ratio?: number;
    }
  ): Promise<void> {
    await this.request(`/issues/${id}.json`, {
      method: "PUT",
      body: JSON.stringify({ issue: data }),
    });
  }

  async addComment(issueId: number, notes: string): Promise<void> {
    await this.request(`/issues/${issueId}.json`, {
      method: "PUT",
      body: JSON.stringify({ issue: { notes } }),
    });
  }

  // ─── Projects ─────────────────────────────────────────────

  async getProjects(params: {
    limit?: number;
    offset?: number;
  } = {}): Promise<{ projects: Project[]; total_count: number }> {
    const query = new URLSearchParams();
    query.set("limit", String(params.limit ?? 25));
    query.set("offset", String(params.offset ?? 0));
    return this.request<{ projects: Project[]; total_count: number }>(
      `/projects.json?${query.toString()}`
    );
  }

  async getProject(id: string): Promise<{ project: Project }> {
    return this.request(`/projects/${id}.json`);
  }

  // ─── Attachments ──────────────────────────────────────────

  async getAttachment(id: number): Promise<{ attachment: Attachment }> {
    return this.request(`/attachments/${id}.json`);
  }
}

File: redmine_mcp_server/redmine_client.py
"""
Redmine REST API クライアント
"""
import httpx
from typing import Any, Optional


class RedmineClient:
    def __init__(self, base_url: str, api_key: str):
        self.base_url = base_url.rstrip("/")
        self.headers = {
            "X-Redmine-API-Key": api_key,
            "Content-Type": "application/json",
        }

    async def _request(
        self,
        method: str,
        path: str,
        params: Optional[dict] = None,
        json: Optional[dict] = None,
    ) -> Any:
        url = f"{self.base_url}{path}"
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.request(
                method, url, headers=self.headers, params=params, json=json
            )
            if response.status_code == 204:
                return None
            if not response.is_success:
                raise Exception(
                    f"Redmine API Error [{response.status_code}]: {response.text}"
                )
            return response.json()

    # ─── Issues ───────────────────────────────────────────────

    async def get_issues(
        self,
        project_id: Optional[str] = None,
        status_id: Optional[str] = None,
        assigned_to_id: Optional[str] = None,
        subject: Optional[str] = None,
        limit: int = 25,
        offset: int = 0,
    ) -> dict:
        params: dict = {"limit": limit, "offset": offset}
        if project_id:
            params["project_id"] = project_id
        if status_id:
            params["status_id"] = status_id
        if assigned_to_id:
            params["assigned_to_id"] = assigned_to_id
        if subject:
            params["subject"] = f"~{subject}"
        return await self._request("GET", "/issues.json", params=params)

    async def get_issue(self, issue_id: int) -> dict:
        return await self._request(
            "GET", f"/issues/{issue_id}.json",
            params={"include": "journals,attachments"}
        )

    async def create_issue(self, data: dict) -> dict:
        return await self._request("POST", "/issues.json", json={"issue": data})

    async def update_issue(self, issue_id: int, data: dict) -> None:
        await self._request("PUT", f"/issues/{issue_id}.json", json={"issue": data})

    async def add_comment(self, issue_id: int, notes: str) -> None:
        await self._request(
            "PUT", f"/issues/{issue_id}.json", json={"issue": {"notes": notes}}
        )

    # ─── Projects ─────────────────────────────────────────────

    async def get_projects(self, limit: int = 25, offset: int = 0) -> dict:
        return await self._request(
            "GET", "/projects.json", params={"limit": limit, "offset": offset}
        )
